### DISCUSSION TITLE
Deploying a model

### DISCUSSION THREAD

**Discussion by assertjohn** [opendilab/DI-engine#508] [Q&A] [answered]
I want to deploy a model in such a way that I can have it observe and act in the real world through an API. It recieves an observation and then it sends back an action. 

My understanding is that the trained model parameters are saved in the .ckpt files, and can be loaded through the config files, however I'm lost after that.

Apologies if this information is already out there, I've looked through all the documentation and wasn't able to figure it out myself. Would really appreciate the help! 

For reference, this is a simplified version of what I'm trying to achieve, but in stable-baselines3:
```
evn = gym.make("my_env")    # load env
model = DQN.load("my_model", env=env)    # load model
env.reset()    # initialize model
obs = get_new_obs()    # get latest observation data from real world with custom function
while True:                                                     
    action, _states = model.predict(obs)    # get latest action
    custom_action(action)    # take an action in the world world with custom function
    obs = get_new_obs()    # get latest observation data from real world with custom function
```

**Comment by PaParaZz1** [comment]
Sorry about there is no an obvious entry in the documentation about how to evaluate (deploy) the trained policy.

You can refer to this [demo](https://github.com/opendilab/DI-engine/blob/main/ding/entry/application_entry.py#L20) and check whether it is enough for you demand.

**Reply by assertjohn** [reply]
Thanks for the reply! I read over the code you linked to and made some progress but I’m still not able to do what I want. 

For reference, I’m working with the gym_anytrading env, which I see has it’s own evaluator class and eval function. 

I have my own scripts for pulling market data and executing trades through APIs, but I’m having a hard time connecting these to the trained model.

I’m happy to pay for help with this if you’re interested, I see you’ve worked with this env recently.

**Comment by PaParaZz1** [comment, marked as answer]
I add a new policy deployment demo like what in SB3. You can refer to [this example](https://github.com/opendilab/DI-engine/blob/main/dizoo/classic_control/cartpole/entry/cartpole_c51_deploy.py).

**Reply by assertjohn** [reply]
this is great, thanks!

**Comment by yourGuy** [comment]
@johnlut were you able to figure it out how to make the demo work for gym_anytrading, I'm a little lost at how to make predictions 
Thank you

**Reply by Cloud-Pku** [reply]
Hello, this is a [demo (colab)](https://colab.research.google.com/drive/1IKfjfyb0PSyTranJemvve2V38_JUs6yB?usp=sharing) for gym-anytrading. If you have any comments, welcome to discuss.

**Reply by yourGuy** [reply]
Thank you that helped a lot. I run it on Forex data which has much smaller fluctuations in price and smaller volumes and the reword seems to converge to 0, any ideas how to tweak the settings to make the model more sensitive to smaller fluctuations. Also is it possible to implement this with the PPO/AS2 models ?
Thank you again

**Reply by Cloud-Pku** [reply]
Can you provide some screenshots of tensorboard and rendering results so that we can check where the problem is?

**Reply by yourGuy** [reply]
![download](https://github.com/opendilab/DI-engine/assets/6990064/382774a9-9863-4bf9-9c3f-100ecb62060a)
![download-1](https://github.com/opendilab/DI-engine/assets/6990064/b505ccc2-59e9-4994-9433-d973abcfef55)
if i print reward form the prediction loop 
```
while True:
        action = forward_fn(obs)
        obs, rew, done, info = env.step(action)
        print(rew)
        returns += rew
        if done:
            break
```
its always [.0]
I think because in training rewards is 0 or very close to 0 
INFO:evaluator_logger:[EVALUATOR]env 0 finish episode, final reward: -0.013898119330406189, current episode: 1
INFO:evaluator_logger:[EVALUATOR]env 1 finish episode, final reward: 0.0, current episode: 2
INFO:evaluator_logger:[EVALUATOR]env 2 finish episode, final reward: 0.0, current episode: 3
INFO:evaluator_logger:[EVALUATOR]env 3 finish episode, final reward: 0.0, current episode: 4
INFO:evaluator_logger:[EVALUATOR]env 4 finish episode, final reward: 0.0, current episode: 5
INFO:evaluator_logger:[EVALUATOR]env 5 finish episode, final reward: 0.0, current episode: 6
INFO:evaluator_logger:[EVALUATOR]env 6 finish episode, final reward: -0.011954807676374912, current episode: 7
INFO:evaluator_logger:[EVALUATOR]env 7 finish episode, final reward: 0.0, current episode: 8
INFO:evaluator_logger:
let me know if you need any other info, all model settings are same as the demo colab just the data is changed to forex data 
Thank you
